### ISSUE TITLE
Hangs on 32-bit platforms

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/UXP#739]
Details: https://forum.palemoon.org/viewtopic.php?p=150281#p150281

> I have been able to reproduce this slowness and hanging issues in Windows 7 x86 virtual machine (2GB RAM, Dual-Core) with no addons installed and just new Pale Mono 28.0.0.1 x86 version. Problem is not present in Windows 7 x64 virtual machine.
> 
> X86 version problems seems rather large. Switching between tabs is very slow and hangs often. Loading websites is slow and can hang. I had 11-13 tabs open in virtual machine during testing.
> 
> I have not been able to reproduce youtube videos not playing problem or crashes, only hangs
> 
> Largest noticeable problems in hang stack of WhatIshang were in stable version of PM: GC_SLICE_MS (telemetry), https://aus5.mozilla.org/update/3/GMP/55.0/%BUILD_ID%/%BUILD_TARGET%/%LOCALE%/%CHANNEL%/%OS_VER, Browser:ForwardOrForwardDuplicate and Gecko_IOThread.
> 
> Largest noticeable problems in hang stack in unstable version: Browser:ForwardOrForwardDuplicate, https://aus5.mozilla.org/update/3/GMP/55.0/%BUILD_ID%/%BUILD_TARGET%/%LOCALE%/%CHANNEL%/%OS_VER and Gecko_IOThread.
> 
> 
> I can't reproduce these hangs in Firefox ESR 52.9 x86 version but I can reproduce them also In Basilisk browser x86 version.
> 
> It doesn't matter if hardware acceleration is force enabled for VM or disabled, hangs happens.

**Comment by wolfbeast** [comment]
Confirmed in a Win7 x86 VM, the browser will start eating up a full core. Opening the browser console shows OOM errors and client-server connect errors in devtools.
Step 1: build a version without devtools, check.

**Comment by JustOff** [comment]
> Opening the browser console shows OOM errors and client-server connect errors in devtools.

Was there any OOM errors if you open the toolkit Error Console instead of the Browser Console?

**Comment by wolfbeast** [comment]
> Were there any OOM errors if you open the toolkit Error Console instead of the Browser Console?

Yup a ton of "uncaught exception: out of memory" errors.
In the toolkit window I'm also seeing the notorious black squares pop up when scrolling.

**Comment by wolfbeast** [comment]
I've been trying a few things so far without success. Th potential candidate of me having made a mistake in removing the crash reporter was also a dud.
Checking basilisk builds, the latest public one also suffers from this issue, but the 2017-12-28 one does not. I think we can at least narrow it down a little by going through the basilisk builds

**Comment by wolfbeast** [comment]
Loks like the very first build on take 2 (24 April) already suffers from this issue. :(

**Comment by JustOff** [comment]
I found a way to reproduce OOM using https://browserbench.org/JetStream/, it can't pass `bigfib.cpp` test and here is what is in the console:

```
"bigfib.cpp: ERROR: https://browserbench.org/JetStream/JetStreamDriver.js:152: uncaught exception: out of memory"  
JetStreamDriver.js:495:9
uncaught exception: out of memory JetStreamDriver.js:152:9
```

But at least Basilisk build from [Functional checkpoint 1](https://github.com/MoonchildProductions/UXP/releases/tag/FullFunction_CP1)  does not seem to have this problem.

**Comment by wolfbeast** [comment]
@adisib actually caught a mistake I made in IPC on 20 Feb which may be related, which is after the functional checkpoint. Since this does seem IPC related, it may fix the issue (here's hoping).

**Comment by win7-7** [comment]
Moonchild you said in freenode chat: ESR 52.6 is unusable in win7 x86 when opening a bunch of tabs concurrently.

Can't reproduce this. Firefox ESR 52.6 does not hang with 11-13 tabs but Basilisk browser does hang so this bug cause might take time to find.

**Comment by JustOff** [comment]
It seems that [Fix incorrect code removal in ipc.](https://github.com/MoonchildProductions/UXP/commit/0bc2511432fe0cdcae92aae71576b6eb9bad7f5c) doesn't help, I still have OOM in `bigfib.cpp@JetStream`.

**Comment by wolfbeast** [comment]
@JustOff no it doesn't help. I just tested it myself.

@win7-7 It pretty much ground to a halt but the behavior was different that whan we're seeing here so whatever bottleneck I ran into isn't the cause of this.

I think the only way to find out which commit caused this is bisection. So somewhere between Feb 20 and April 24 this started being an issue. Anyone feel up to the task?

**Comment by JustOff** [comment]
I can participate in the bisection, but my PC requires ~100 minutes to build Basilisk :poop:

**Comment by wolfbeast** [comment]
You don't necessarily have to clobber each time ;)
For me I have a fast build but it's annoying to get the binaries to the test environment from the external machine, so that takes time, too. I'll see how far I get by myself tomorrow.

**Comment by JustOff** [comment]
I have an archive of Basilisk builds from [NSS bisection](https://github.com/MoonchildProductions/UXP/issues/265#issuecomment-384255120), I will now try to check them and report back.

**Comment by JustOff** [comment]
This allowed me to narrow the window to 03/29 - 04/23.

**Comment by JustOff** [comment]
New window is 04/15 - 04/19, and now it's time to sleep)

**Comment by wolfbeast** [comment]
Confirmed the performance issues are part of the [0b5cb0859]-[9135a11e8] range of upstream commits.
The jetstream error is likely due to the commit indicated but that won't impact the performance issues we've seen as it is exclusive to asm.js -- more likely is the impact of [e7189e33f] for that, but further builds will have to give clarity there.

**Comment by JustOff** [comment]
It's great that you could confirm the range specified above, but then I'm surprised that it coincided with what I found for asm.js exclusively. Unfortunately, I still have not found how to reproduce the general performance issue in my environment, so that the further bisection remains for you.

**Comment by wolfbeast** [comment]
OK, I've built 3 versions of Basilisk, one at [d66d343e1], one at [e7189e33f] and one at [947213627] (74204, 80743 and 90818, respectively). My own experiments show:
Build | Perf | Test
----- | ---- | ----
74204 | OK | OK
80743 | ? | OK
90818 | Bad | Bad

I'm not entirely sure about the perf, it seems all over the place but 80743 doesn't seem to choke on shutdown with max CPU while 90818 does. 
Uploaded the builds to ftp://publicbeta:get@ftp.palemoon.org/trybuilds/ -- other people please verify?

**Comment by wolfbeast** [comment]
Right, that's about what I saw. So the main culprit is 1388020, and made worse by 1444668. I'm building Pale Moon unstable with both backed out now and will upload it to the same location when done.

**Comment by JustOff** [comment]
What is interesting, that ESR 52.7/52.8/52.9 do not have problem with JetStream (with and without e10s).

**Comment by win7-7** [comment]
I would say problem solved.

By the way ftp://ftp.palemoon.org/trybuilds/ produces 550 CWD failed, directory not found.

I managed to get this unstable build before problem started with ftp just while ago.

**Comment by JustOff** [comment]
> I would say problem solved.

It would be nice to confirm this using Pale Moon unstable when it's published.

> By the way ftp://ftp.palemoon.org/trybuilds/ produces 550 CWD failed, directory not found.

You have to use credentials, i.e. ftp://publicbeta:get@ftp.palemoon.org/trybuilds/

**Comment by win7-7** [comment]
No problem with ftp. Just pasted accidentally ftp://ftp.palemoon.org/trybuilds/ instead of ftp://publicbeta:get@ftp.palemoon.org/trybuilds/.

**Comment by satrow** [comment]
Initial test on W7x64 indicates a much more compressed CPU/network graph when loading a saved ~50 tab session cf. 28.0.0.1 the first time, ~50% faster on a slightly slower (~4.5 Mbps from 5 yesterday) connection; my 'solo' time slot's just finished, too much shared traffic to do another comparison, will continue to test for disconnects.

**Comment by mattatobin** [comment]
I am gonna test Win64 builds.. As I learn more about the full nature of what is happening there may be a few instances where this has had far less but noticable impact.

In any case i should check Win64 in general anyway.

**Comment by JustOff** [comment]
@wolfbeast, could you please clarify a bit how these reverts affect security?

**Comment by wolfbeast** [comment]
@JustOff They don't affect security.

One was in response to a fuzzer-sourced overflow that in practice can't be triggered from web content, and even if so it would only affect an e10s browser's security (potential sandbox escape from a corrupted content process).

The other was in response to a potential overflow for the WASM assembler. WASM is disabled by default, for starters, and it requires a very specific set of circumstances to happen.

**Comment by mattatobin** [comment]
tl;dr defense in depth had to give way to real world concerns.

**Comment by wolfbeast** [comment]
@trav90 I'm planning on making this a point release. The backed-out DiD bugs' performance impact is likely going to be cross platform. If you can make builds for Linux tonight i can push the update Friday.

**Comment by trav90** [comment]
Yup I’m already building them now. :)

**Comment by trav90** [comment]
Linux versions are built, signed, and ready for release Friday.

**Comment by win7-7** [comment]
On a related matter could you restore previously used compiler optimizations soon after this now that hangs on x86 have been resolved and confirmed to not be caused by compiler optimizations: Qpar -Qpar-report:1 -Qfast_transcendentals? It would be good to have more optimized browser.

**Comment by JustOff** [comment]
AFAIU, "[Switch to using a single memory allocation arena](https://github.com/MoonchildProductions/UXP/commit/4ceb21241eacac2911f2fed846359215870f121f)" was another attempt to mitigate the same issue. Do we still need these changes after all?
